test/integration/multinode_test.go

```diff
@@ -42,269 +50,280 @@ func TestMultiNode(t *testing.T) {
 			validator validatorFunc
 		}{
 			{"FreshStart2Nodes", validateMultiNodeStart},
+			// Add worker node
 			{"AddNode", validateAddNodeToMultiNode},
-			{"StopNode", validateStopRunningNode},
-			{"StartAfterStop", validateStartNodeAfterStop},
-			{"DeleteNode", validateDeleteNodeFromMultiNode},
+			{"StopNode", validateStopRunningNode(ThirdNodeName)},
+			{"StartAfterStop", validateStartNodeAfterStop(ThirdNodeName)},
+			{"DeleteNode", validateDeleteNodeFromMultiNode(ThirdNodeName, true)},
+			// Add control plane node
+			{"AddControlPlaneNode", validateAddControlPlaneNodeToMultiNode},
+			{"StopControlPlaneNode", validateStopRunningNode(ThirdNodeName)},
+			{"StartControlPlaneAfterStop", validateStartNodeAfterStop(ThirdNodeName)},
+			{"DeleteControlPlaneNode", validateDeleteNodeFromMultiNode(ThirdNodeName, true)},
+			// Test cluster stop && start
 			{"StopMultiNode", validateStopMultiNodeCluster},
 			{"RestartMultiNode", validateRestartMultiNodeCluster},
 		}
+
+		s := &clusterStatus{}
+
+		if DockerDriver() {
+			rr, err := Run(t, exec.Command("docker", "version", "-f", "{{.Server.Version}}"))
+			if err != nil {
+				t.Fatalf("docker is broken: %v", err)
+			}
+			if strings.Contains(rr.Stdout.String(), "azure") {
+				s.isAzure = true
+			}
+		}
+
 		for _, tc := range tests {
 			tc := tc
 			if ctx.Err() == context.DeadlineExceeded {
 				t.Fatalf("Unable to run more tests (deadline exceeded)")
 			}
 			t.Run(tc.name, func(t *testing.T) {
 				defer PostMortemLogs(t, profile)
-				tc.validator(ctx, t, profile)
+				tc.validator(ctx, t, profile, s)
 			})
 		}
 	})
 }
 
-func validateMultiNodeStart(ctx context.Context, t *testing.T, profile string) {
+func validateMultiNodeStart(ctx context.Context, t *testing.T, profile string, s *clusterStatus) {
 	// Start a 2 node cluster with the --nodes param
 	startArgs := append([]string{"start", "-p", profile, "--wait=true", "--memory=2200", "--nodes=2", "-v=8", "--alsologtostderr"}, StartArgs()...)
 	rr, err := Run(t, exec.CommandContext(ctx, Target(), startArgs...))
 	if err != nil {
 		t.Fatalf("failed to start cluster. args %q : %v", rr.Command(), err)
 	}
 
-	// Make sure minikube status shows 2 nodes
-	rr, err = Run(t, exec.CommandContext(ctx, Target(), "-p", profile, "status", "--alsologtostderr"))
-	if err != nil {
-		t.Fatalf("failed to run minikube status. args %q : %v", rr.Command(), err)
-	}
-
-	if strings.Count(rr.Stdout.String(), "host: Running") != 2 {
-		t.Errorf("status says both hosts are not running: args %q: %v", rr.Command(), rr.Stdout.String())
-	}
-
-	if strings.Count(rr.Stdout.String(), "kubelet: Running") != 2 {
-		t.Errorf("status says both kubelets are not running: args %q: %v", rr.Command(), rr.Stdout.String())
-	}
-
+	s.startCluster()
```

Comment:
First, the implementation code looks great. My only hesitation is with this integration test: Could you please remove these helper methods and in-line the test code instead? 

Readers should need to read only the test body to understand the test. My advice is not sacrifice test clarity in the interest of removing code duplication. People only read tests when they are broken, so lets make them as simple as possible.

I would also suggest removing the *clusterStatus argument as well, if possible.

Thank you so much for your contribution! I can't wait to see it merged.

Reply:
IMO helper functions will make things more clear. In this situation, it's hard to set these constants right, especially they affect each other across so many functions.
What's more, we need a status variable to make functions like `stop/start/delete` reusable.

Reply:
/ping
what's your opinion @medyagh